papis/commands/tag.py

```diff
@@ -0,0 +1,148 @@
+
+    .. code:: sh
+
+        papis tag --remove TAG1 QUERY
+
+  This removes TAG1 from the document. If it didn't exist before, nothing 
+  happens. You can use `--remove` as many times as you like to remove multiple
+  tags.
+
+- Use `--clear` (or `-c`) to remove all tags:
+
+    .. code:: sh
+
+        papis tag --clear --new TAG1 QUERY
+
+  This removes all tags from the document and then adds TAG1. You can of course
+  also use `--clear` by itself to simply remove all tags without adding new 
+  ones.
+
+- Use `--all` (or `-a`) with any of the other options to apply the tagging
+  operation to all matching documents:
+
+    .. code:: sh
+
+        papis tag --all --new TAG1 QUERY
+  
+  This adds TAG1 to all documents matching the QUERY rather than opening the
+  picker to let you choose one.
+
+"""
+from typing import Tuple, Set, List
+
+import click
+
+import papis.config
+import papis.commands
+import papis.document
+import papis.strings
+import papis.cli
+
+logger = papis.logging.get_logger(__name__)
+
+TAG_FORMAT = papis.config.get("tag-format")
+
+
+def make_set_of_tags(tags: List[str] | str) -> Set[str]:
+    if isinstance(tags, list) and TAG_FORMAT == "list":
+        set_of_tags = set(tags)
+    elif isinstance(tags, str) and TAG_FORMAT == "comma_separated":
+        set_of_tags = set(tags.split(", "))
+    elif isinstance(tags, str) and TAG_FORMAT == "semicolon_separated":
+        set_of_tags = set(tags.split("; "))
+    elif isinstance(tags, str) and TAG_FORMAT == "space_separated":
+        set_of_tags = set(tags.split(" "))
+    else:
+        logger.error(
+            '`tag-format` is set to "%s", but we got a %s. '
+            + "Please configure `tag-format` correctly or fix the info.yaml",
+            TAG_FORMAT,
+            type(tags).__name__,
+        )
+        raise SystemExit(1)
+    return set_of_tags
+
+
+def convert_to_tag_format(tags: Set[str]) -> str | List[str]:
+    if TAG_FORMAT == "list":
+        converted_tags = list(tags)
+    elif TAG_FORMAT == "comma_separated":
+        converted_tags = ", ".join(tags)
+    elif TAG_FORMAT == "semicolon_separated":
+        converted_tags = "; ".join(tags)
+    else:
+        converted_tags = " ".join(tags)
+    return converted_tags
+
+
+@click.command("tag")
+@click.help_option("--help", "-h")
+@click.option("--new", "-n", multiple=True, help="Add a tag")
+@click.option("--remove", "-r", multiple=True, help="Remove a tag")
+@click.option("--clear", "-c", is_flag=True, help="Clear all tags")
+@papis.cli.doc_folder_option()
+@papis.cli.all_option()
+@papis.cli.query_argument()
+def cli(
+    new: str,
+    remove: str,
+    clear: bool,
+    query: str,
+    doc_folder: Tuple[str, ...],
+    _all: bool,
+) -> None:
+    """
+    Add and delete document tags.
+    """
+
+    if not TAG_FORMAT:
+        logger.warning(
+            "The `tag-format` option isn't set. Please define it in your config."
+        )
+        return
```

Comment:
The reason why I didn't do this is that we might then end up adding tags in a wrong format. For instance, let's say the default is a list (which I think would be the most sensible). The user, however, is using comma separated strings (but haven't specified that in the config). If they then add a tag to a document that hasn't any tags yet, no error will be generated (because an error requires that `make_set_of_tags()` finds a string). We then add a list item to the doc, leading to inconsistencies. Do you have some idea of how to deal with this?

Reply:
Yeah, haven't thought of it too well. Like you say, the main issue is that the user may have tags in various formats because we haven't standardized them in any way so it's a wild wild west. Maybe we can add a big warning at the beginning on first use or something?
```
logger.warning(
	"This command assumes that your tags are separated by {separator}. If this is not correct, 
	it will not split them correctly and may give unexpected results. 
	Use `papis doctor -t normalize-tags --fix` to clean up.")
papis.tui.utils.confirm("Are you sure you want to continue?")
```

Will have to think about it some more, so for now feel free to leave it as is!

Reply:
Yeah, a warning like that would be good, too. Ok, for now I'll leave it, we can easily change it later.